staging/src/kubevirt.io/client-go/api/v1/schema.go

```diff
@@ -201,17 +201,19 @@ type EFI struct {
 // +k8s:openapi-gen=true
 type ResourceRequirements struct {
 	// Requests is a description of the initial vmi resources.
-	// Valid resource keys are "memory" and "cpu".
+	// Valid resource keys are "memory", "memory-overhead" and "cpu".
 	// +optional
 	Requests v1.ResourceList `json:"requests,omitempty"`
 	// Limits describes the maximum amount of compute resources allowed.
-	// Valid resource keys are "memory" and "cpu".
+	// Valid resource keys are "memory", "memory-overhead" and "cpu".
 	// +optional
 	Limits v1.ResourceList `json:"limits,omitempty"`
-	// Don't ask the scheduler to take the guest-management overhead into account. Instead
-	// put the overhead only into the container's memory limit. This can lead to crashes if
-	// all memory is in use on a node. Defaults to false.
-	OvercommitGuestOverhead bool `json:"overcommitGuestOverhead,omitempty"`
+	// KubevirtMemoryOverhead describes the memory overhead that results from the libraries
+	// and other dependencies used by the Kubevirt platform. It applies when "memory-overhead"
+	// is not specified as part of Limits and therefore determined by the system.
+	// If omitted, the cluster-level setting applies.
+	// +optional
+	KubevirtMemoryOverhead *resource.Quantity `json:"platformMemoryOverhead,omitempty"`
```

Comment:
Isn't this now completely redundant with the introduction of `memory-overhead`?

Reply:
No - it is not used only when `memory-overhead` is specified. When memory-overhead is not specified, we compute it as before, and this property is then taken into account in that calculation

Reply:
I see. I have the impression that things got now even more convoluted and complicated than before. I could before already set more memory in requests and limits and setting a value for the VMI in the domain section. That is sufficient for sidecars and has a similar effect.

Reply:
Well, it depends on the perspective we look at it from I guess.
Yes, it adds more configuration properties that may look more convoluted and complicated, I agree.
But I think that nevertheless it improves the API:

As an openshift user I define the requests and limits per container for all relevant containers that are comparised by my pod, including sidecars, right? I say how much my workload should consume and how much the "infrastrcuture" level should consume.

But as a Kubevirt adminitrator I have this weird entity "VMI" that abstracts the lower level details for me, that the pod is derived from, yet when I configure the VMI I need to think about the lower-level infrastructure - either the overhead introduced by the infrastructure stack (e.g., libvirt) or sidecar containers. This makes sense to us since we currently breath openshift, but is really non-trivial way of thinking about the VMI (in my opinion).

I think that if we're able to separate out the workload related, i.e., the VM stuff, from the infrastructure level by separate properties, the API would be more virtualization-friendly.

I don't want to drag it into a discussion on subjective things though - it can be an endless discussion :)

So more concretely, I think that once we enable users to specify the infrastructure related things separately, we can then change the semantics of Limits.Memory in Kubevirt. I really think it doesn't fit VMs even when memory hot-plug is not supported.

Reply:
> But as a Kubevirt adminitrator I have this weird entity "VMI" that abstracts the lower level details for me, that the pod is derived from, yet when I configure the VMI I need to think about the lower-level infrastructure - either the overhead introduced by the infrastructure stack (e.g., libvirt) or sidecar containers. This makes sense to us since we currently breath openshift, but is really non-trivial way of thinking about the VMI (in my opinion).

You only have to think about this when you introduce sidecar containers or if we do something wrong. It is something you should not have to even think about normally (not as admin, nor as ordinary user).

> So more concretely, I think that once we enable users to specify the infrastructure related things separately, we can then change the semantics of Limits.Memory in Kubevirt. I really think it doesn't fit VMs even when memory hot-plug is not supported.

Users normally don't have to specify anything like this and don't have to think about it. With the changes the internal overhead is much more prominent and everyone sees it now always, no matter if it has any meaning for you or not. Why would we want to change limits.memory? As soon as we have ballooning it makes even more sense the k8s way.

Reply:
> You only have to think about this when you introduce sidecar containers or if we do something wrong. It is something you should not have to even think about normally (not as admin, nor as ordinary user).

Right, that's always the easir case - same with the additional configuration properties introduced in the PR, users would simply ignore them. 
The question is about the other case, when users do need to set them, whether they should (1) compute the Requests.Memory and Limits.Memory accordingly; or (2) specify them separately via different properties.
 
> > So more concretely, I think that once we enable users to specify the infrastructure related things separately, we can then change the semantics of Limits.Memory in Kubevirt. I really think it doesn't fit VMs even when memory hot-plug is not supported.
> 
> Users normally don't have to specify anything like this and don't have to think about it. With the changes the internal overhead is much more prominent and everyone sees it now always, no matter if it has any meaning for you or not. Why would we want to change limits.memory? As soon as we have ballooning it makes even more sense the k8s way.

As for the first part, yeah, it is exposed to users but in my opinion it does make sense to reflect those changes - I requests 128M of memory but got more, so I may want to know that the system altered the configuration I provided at the VMI level. That's debatable though, I agree.

As for the second part, imagine when supporting hot-plugging - I increases the memory consumed by the guest from 8G to 16G and the VM started with limit of 10G. It does make sense to limit the amount of memory to be plugged, but it doesn't make sense for the OOM killer to kill it because of that :)
I already voiced the following argument on the mailing list - on k8s it makes a lot of sense to protect the system against pods that consume too much memory (that can happen due to extreme load, bugs, etc). But in Kubevirt, the workloads would run inside a VM whose memory is bounded. Such memory limits only protect against bugs in the virtualization stack, which is considered redundant [1].

[1] To protect against sidecar containers consuming too much memory, I think the limit should be specified at the container level, not the VMI.

Reply:
> I already voiced the following argument on the mailing list - on k8s it makes a lot of sense to protect the system against pods that consume too much memory (that can happen due to extreme load, bugs, etc). But in Kubevirt, the workloads would run inside a VM whose memory is bounded. Such memory limits only protect against bugs in the virtualization stack, which is considered redundant [1].

You can already right now just set the memory.limit, this gives you guaranteed QoS and the VMI will have the memory of limits ( the limits on the pod is what you set + overhead). If you want burstable, you will just set memory.requests, or set memory.requests to something lower (here we have the only thing which should be fixed, we should use always memory.limts if it set and no domain memory is requested).

Later when we have ballooning these two values will exactly represent what you expect: requests is minimum memory, limit is what the VM can at max have (yes, it should not exceed it because it can't go above it). 

> [1] To protect against sidecar containers consuming too much memory, I think the limit should be specified at the container level, not the VMI.

The sidecars have potential effects on qemu/libvirt by modifying libvirt domain xml, which can increase the memory-needs.


I think that we can add/change/rmove fields here and at the end it will still not be clearer.

Reply:
> > I already voiced the following argument on the mailing list - on k8s it makes a lot of sense to protect the system against pods that consume too much memory (that can happen due to extreme load, bugs, etc). But in Kubevirt, the workloads would run inside a VM whose memory is bounded. Such memory limits only protect against bugs in the virtualization stack, which is considered redundant [1].
> 
> You can already right now just set the memory.limit, this gives you guaranteed QoS and the VMI will have the memory of limits ( the limits on the pod is what you set + overhead). If you want burstable, you will just set memory.requests, or set memory.requests to something lower (here we have the only thing which should be fixed, we should use always memory.limts if it set and no domain memory is requested).

I think we talk about different things here, or I don't fully understand.
I agree that the current API is very flexible and enables doing all these kinds of things but I argue that specifically, the limit.memory with its current semantics is not so useful with virtualization. I do agree that behind the scenes we'll need to default it properly to achieve things like the desired QoS due to the infrastructure we use.

> 
> Later when we have ballooning these two values will exactly represent what you expect: requests is minimum memory, limit is what the VM can at max have (yes, it should not exceed it because it can't go above it).

I don't kow, in both cases (when using ballooning or not) the scheduler assures that the minimum guaranteed memory is reserved on the host and the limit is what the VM can possibly consume.

Why do you find the memory balloon functionality having a major effect on the semantics of those settings (besides the optimization of passing memory back to the host)?

> 
> > [1] To protect against sidecar containers consuming too much memory, I think the limit should be specified at the container level, not the VMI.
> 
> The sidecars have potential effects on qemu/libvirt by modifying libvirt domain xml, which can increase the memory-needs.
> 
> I think that we can add/change/rmove fields here and at the end it will still not be clearer.

Right, that's why I want to avoid this discussion to get to the "my API is nicer/clearer/simpler" kind of arguments.

From the functional point of view it enables:
1. To override the memory overhead completely per-VMI by specifying it in requests.memory-overhead and limits.memory-overhead rather than by setting OvercommitGuestOverhead=true and changing computing the values of limits.memory and requests.memory
2. It enables keeping all the existing stuff but to override the platform's memory overhead at the cluster and VM level, which is currently hard-coded

Reply:
> Right, that's why I want to avoid this discussion to get to the "my API is nicer/clearer/simpler" kind of arguments.

The issue is that the new fields don't add new possibilities to the already existing ones.

> From the functional point of view it enables:
> 
> 1. To override the memory overhead completely per-VMI by specifying it in requests.memory-overhead and limits.memory-overhead rather than by setting OvercommitGuestOverhead=true and changing computing the values of limits.memory and requests.memory

you don't have to set OvercommitGuestOverhead. You can do all fine-tuning for additional overheads with `memory.limits`, `memory.requests` and `spec.domain.memory.guest`.

For instance setting `spec.domain.memory.guest` to 1000Mi and `memory.requests` to `1100Mi` will add 100 Mi additional requested memory for e.g. sidecar-caused side-effects to the normal overhead calculation.

> 2. It enables keeping all the existing stuff but to override the platform's memory overhead at the cluster and VM level, which is currently hard-coded

Not debating the cluster-level setting. Need to look again into the details there, but it works for me having it on the cluster level.

Reply:
> > Right, that's why I want to avoid this discussion to get to the "my API is nicer/clearer/simpler" kind of arguments.
> 
> The issue is that the new fields don't add new possibilities to the already existing ones.

Let's say I have a VM with 1G of virtual memory, cluster overcommit=200% and I want to change the memory overhead of this VMI to 256M.

With the current API I need to:
1. Compute the memory-request that is derived from the virtual memory+overcommit ratio (512M)
2. Figure out the default memory overhead in kubevirt
3. Compute the detla between the desired memory overhead and the above
4. Add the result of (3) with that of (1)
5. Set the result of (4) on the VMI spec

Alternatively, with the new fields I simply change the memory overhead field to 256M.
Not only it requires less computations on the client side, but it is much closer to what I need.

I agree that there it doesn't enable users to define something they cannot achieve with the current API but I do think it adds new possibilities in terms of usability.

> 
> > From the functional point of view it enables:
> > 
> > 1. To override the memory overhead completely per-VMI by specifying it in requests.memory-overhead and limits.memory-overhead rather than by setting OvercommitGuestOverhead=true and changing computing the values of limits.memory and requests.memory
> 
> you don't have to set OvercommitGuestOverhead. You can do all fine-tuning for additional overheads with `memory.limits`, `memory.requests` and `spec.domain.memory.guest`.
> 
> For instance setting `spec.domain.memory.guest` to 1000Mi and `memory.requests` to `1100Mi` will add 100 Mi additional requested memory for e.g. sidecar-caused side-effects to the normal overhead calculation.

Indeed.
I had the flow I mentioned above in mind but I think (2) is problematic - I can ask on the mailing list what the memory overhead would be or look at the code, but that's too cumbersome. So I guess that in this case I would have said "Gosh, it's too difficult to understand what is going on so I'll just increase the entire memory overhead to 512M - that should do it". For this, I would need to make sure what kubevirt computes doesn't get into the requested memory (therefore setting OvercommitGuestOverhead to true) and then add 512M to the result of (1).

> 
> > 1. It enables keeping all the existing stuff but to override the platform's memory overhead at the cluster and VM level, which is currently hard-coded
> 
> Not debating the cluster-level setting. Need to look again into the details there, but it works for me having it on the cluster level.

Ah ok, I thought you requested it to be at the VMI level before.